### PR TITLE
Some BIIIG api refactorings

### PR DESCRIPTION
**Pull request by algernon**
This is built on top of #20, because that made both the refactoring, and #20 itself easier (for me, at least). The refactor itself is in d64ce49ae7bca53fbab72c8f1734118844cd37fe, in a single commit.

It implements something very similar to the idea outlined in #29, while introducing no new features, nor regressions (as far as I can tell: the integration tests still pass).

Further commits (b155e759f4ef6aba3f0a3c544cd50d2e22d2daf9 and 34c770a3ee8791261c3a34711e39859d043e66c2) implement additional optimizations, and minor new features: the first makes `/v1/codeberg` more performant, getting rid of the redirect. The second introduces `/v1/sr.ht` and `/v1/git.sr.ht`, where both use the SourceHut's flagship instance, `git.sr.ht`.

I'd suggest checking the branch out, rather than looking at the diff. The diff moves so much stuff around that its hard to follow. The whole thing starts with `src/api/v1/forge.rs`, and each forge is implemented in `src/api/v1/forges/<forge>.rs`. There are a few helper modules here and there, but these are the main entrypoints, I think.

### CODE REVIEW

**Comment by algernon** [comment]
Oh, wait: there *is* one new feature introduced by this: `/v1/flakehub/:user/:repo/tag/:version` and `/v1/flakehub/:user/:repo/t/:version` are now valid endpoints (previously they weren't implemented), because both the `/tag` and `/version` endpoints use the same `get_tarball_url_for_version` handler.

**Comment by algernon** [comment]
Going back to draft, because I identified a few more things to improve about this.

**Comment by algernon** [comment]
Okay. I *think* this is in its final form now.

**Comment by algernon** [comment]
Rebased on top of main, and I took the liberty of adjusting the integration tests so that the codeberg check is no longer allowed to fail. (they all pass)

**Comment by algernon** [comment]
FWIW, I have a few ideas about how to improve this further, but this is already a large PR, so I'll submit them separately. They're considerably smaller in scale:

One idea is to lift out `get_routes()` and `get_selfhosted_routes()` from the trait (also making `new()` in the trait unnecessary), and move them to `api::v1::routes` instead, taking a parameter that implements the `Forge` trait.

...and now that I started to write this down, I realized that the second idea is related to #31, which will need a rebasing anyway, and I can implement the idea there, because it is related to semver anyway. The goal there is to make it possible for forge implementations to skip the "grab a list of releases" step for the semver endpoint, if the forge already has suitable semver support. In practice, this means that for Flakehub, the `/semver/:version` endpoint would just quickly redirect to `flakehub.com/f/<user>/<repo>/<version>`, pretty much as-is, without the need to grab a release list and do the filtering ourselves.

**Review by cafkafk — APPROVED** [review]
I like this refactor a lot, glad to see the introduction of ForgeError rather than just copy pasting that one string around :p
